geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java

```diff
@@ -377,6 +382,25 @@ long zrevrank(byte[] member) {
     return scoreSet.size() - scoreSet.indexOf(orderedSetEntry) - 1;
   }
 
+  ImmutablePair<Integer, List<byte[]>> zscan(Pattern matchPattern, int count, int cursor) {
+    // No need to allocate more space than it's possible to use given the size of the sorted set. We
+    // need to add 1 to zcard() to ensure that if count > members.size(), we return a cursor of 0
+    long maximumCapacity = 2L * Math.min(count, zcard() + 1);
+    if (maximumCapacity > Integer.MAX_VALUE) {
+      LogService.getLogger().info(
+          "The size of the data to be returned by zscan, {}, exceeds the maximum capacity of an array. A value for the ZSCAN COUNT argument less than {} should be used",
+          maximumCapacity, Integer.MAX_VALUE / 2);
+      throw new IllegalArgumentException("Requested array size exceeds VM limit");
+    }
+    List<byte[]> resultList = new ArrayList<>((int) maximumCapacity);
+    do {
+      cursor = members.scan(cursor, 1,
```

Comment:
Something I find out about our scan code is that we implemented SizeableObject2ObjectOpenCustomHashMapWithCursor#scan to take a "count" parameter that allows it to loop but then when we call it we always give it "1" and the caller does a loop. It seems like we could refactor it so that the shared scan method on our HashMap does all the looping.

Reply:
Good call. It seems that all the tests pass with this change, so I've applied it to simplify things.

Reply:
I think this breaks things a bit. @DonalEvans the assertion that was removed in `RedisHashTest.hscanOnlyReturnsElementsMatchingPattern` should remain. If that test is translated into an integration test you'll see it passes in native redis (with a cursor result of `0`).
The javadoc for `SizeableBytes2ObjectOpenCustomHashMapWithCursor.scan` is slightly misleading (for me at least) in that it says:
> This method will scan until at least count entries are returned...
It should preferably say: _"until at least count entries are consumed"_. The caller of this method,`RedisHash.hscan`, has a different semantic for `count` in that it wants to _return_ that many entries that _match_ the pattern not just process that many entries in the hash.

Reply:
This is the integration test I tried:
```
  @Test
  public void wat() {
    Map<String, String> hashy = new HashMap<>();
    hashy.put("ak1", "v1");
    hashy.put("k2", "v2");
    hashy.put("ak3", "v3");
    hashy.put("k4", "v4");
    jedis.hset(HASH_KEY, hashy);

    ScanResult<Map.Entry<byte[], byte[]>> result =
        jedis.hscan(HASH_KEY.getBytes(), "0".getBytes(), new ScanParams().count(3).match("a*"));

    assertThat(result.getCursor()).isEqualTo("0");
  }
```

Reply:
The difficulty here is that Redis has different behaviour depending on the size of the sorted set/hash. For sizes less than a certain amount, they internally use a different data structure, which results in different behaviour:

>"When iterating Sets encoded as intsets (small sets composed of just integers), or Hashes and Sorted Sets encoded as ziplists (small hashes and sets composed of small individual values), usually all the elements are returned in the first SCAN call regardless of the COUNT value."

from https://redis.io/commands/scan#the-count-option

But for larger sizes, a scan command may return no results (except a cursor value) when used with MATCH and COUNT. I created a hash with ~500 entries, three of which matched the pattern `a*`, and called HSCAN with `MATCH a*` and various COUNT values. Redis returned an empty array for COUNT values up to 106, and only one result for COUNT values up to 243, so it's not accurate to say that the COUNT argument indicates the number of matching entries that should be returned.

Given the inconsistent behaviour of Redis' scan commands, I'm not sure that we want to try to exactly match behaviour for situations where a full iteration isn't guaranteed, because our internal implementation will not behave the same for small set/hash sizes. This can be tested by creating a hash with 20 entries, then calling HSCAN with a COUNT value of 5. Redis will return all entries, but our implementation will return 5.

Reply:
That's pretty annoying :) I'd still argue for restoring the prior behavior (or some variant) since it should be more performant overall in that it will reduce the potential number of client calls. If we only process `count` entries for every `scan` call, the client is going to make ~`set_size / count` requests until they are done. The downside of the original implementation is that the key is locked for the duration of the call as more entries are iterated. Since the semantics are so loose, it would be awesome if the scan could be interrupted by a write operation, but that's just crazy talk...

Even Redis requires a lot of calls when the matches are sparse. With a set of 10003 entries where 3 entries will match the `zscan` call (and count=3), Redis still takes 2769 calls before the whole set is processed, so I guess leaving this change is OK for now.

Reply:
Yeah the COUNT/MATCH interaction isn't ideal. From Redis' SCAN documentation:
> It is important to note that the MATCH filter is applied after elements are retrieved from the collection, just before returning data to the client. This means that if the pattern matches very little elements inside the collection, SCAN will likely return no elements in most iterations.

so it'd be a significant departure from Redis' way of doing things if we kept the old implementation (even if it's better).